src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -182,191 +182,130 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
 										 ALLOCSET_DEFAULT_MAXSIZE);
 	oldContext = MemoryContextSwitchTo(localContext);
 
-	/* find transactions that were committed, but not yet confirmed */
-	unconfirmedTransactionList = UnconfirmedWorkerTransactionsList(groupId);
-	unconfirmedTransactionList = SortList(unconfirmedTransactionList, CompareNames);
+	/* take table lock first to avoid running concurrently */
```

Comment:
> Otherwise, it seems possible that we might commit an ongoing distributed transaction that has started and prepared after ActiveDistributedTransactionNumbers() and before PendingWorkerTransactionList() called?

Good point. This requires some more care.

Consider the following order:

1) P = PendingWorkerTransactionList()
2) A = ActiveDistributedTransactionNumbers() 
3) T = systable_beginscan(pgDistTransaction) - get snapshot of pg_dist_transaction

At time 2, we get a conclusive answer to which distributed transactions in P are still in progress. We therefore will never roll forward prepared transactions of in progress transactions.

However, a new transaction could have started, left prepared transactions, and committed between 1 and 2 (or 1 and 3). We wouldn't roll these prepared transactions forward since they are not in P, but we would delete their records, meaning that we would incorrectly roll them back the next time recover_prepared_transactions is called.

A solution is to get the list of prepared transactions again after obtaining the pg_dist_transaction snapshot:

4) Q = PendingWorkerTransactionList()

We then skip deletion for records that appear in Q and we resolve prepared transactions in P.

Does this make sense? Basically, anything that's in P and Q, but not in A, is known to have prepared before 1 and aborted/committed before 2, and is therefore safe to resolve after 2.

Reply:
Actually, I observed such cases on my tests:

```SQL
 SELECT recover_prepared_transactions();
WARNING:  prepared transaction with identifier "citus_0_21505_69107_948" does not exist

```

Reply:
Good catch. Changed to always skipping prepared transactions in Q.

> P = prepared transactions on worker
> (One of the transactions commits in between)
> A = active distributed transactions (doesn't include that transaction)

Either that transaction leaves prepared transactions, which will show up in Q and we skip, or it commits them and it won't show up in Q, in which case deleting the record is fine.